File: docuflow_backend/src/main/java/com/example/docuflow_backend/service/DocumentService.java
package com.example.docuflow_backend.service;

import com.example.docuflow_backend.dto.DocumentDTO;
import com.example.docuflow_backend.model.DocumentStatus;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface DocumentService {
    
    DocumentDTO uploadDocument(MultipartFile file, String title, String description, String username);
    
    DocumentDTO getDocument(Long id);
    
    byte[] getDocumentContent(Long id);
    
    List<DocumentDTO> getAllDocuments();
    
    List<DocumentDTO> getDocumentsByUser(String username);
    
    List<DocumentDTO> getDocumentsByStatus(DocumentStatus status);
    
    List<DocumentDTO> searchDocuments(String query);
    
    DocumentDTO updateDocumentMetadata(Long id, String title, String description, String username);
    
    DocumentDTO updateDocumentStatus(Long id, DocumentStatus newStatus, String username, String comments);
    
    void deleteDocument(Long id);
}

File: docuflow_frontend/src/services/DocumentService.js
import api from "./api"

const DOCUMENTS_URL = "/documents"

export const documentService = {
  // Get all documents
    getAllDocuments: async () => {
        const response = await api.get(DOCUMENTS_URL)
        return response.data
    },

    // Get document by ID
    getDocument: async (id) => {
        const response = await api.get(`${DOCUMENTS_URL}/${id}`)
        return response.data
    },

    // Get document content (file download)
    getDocumentContent: async (id) => {
        const response = await api.get(`${DOCUMENTS_URL}/${id}/content`, {
            responseType: "blob",
        })
        return response.data
    },

    // Get documents by user
    getDocumentsByUser: async (username) => {
        const response = await api.get(`${DOCUMENTS_URL}/user/${username}`)
        return response.data
    },  

    // Get documents by status
    getDocumentsByStatus: async (status) => {
        const response = await api.get(`${DOCUMENTS_URL}/status/${status}`)
        return response.data
    },  

    // Search documents
    searchDocuments: async (query) => {
        const response = await api.get(`${DOCUMENTS_URL}/search`, {
            params: { query },
        })
        return response.data
    },

    // Upload document
    uploadDocument: async (file, title, description) => {
        const formData = new FormData()
        formData.append("file", file)
        formData.append("title", title)
    
        if (description) {
            formData.append("description", description)
        }
    
        const response = await api.post(DOCUMENTS_URL, formData, {
            headers: {
                "Content-Type": "multipart/form-data",
            },
        })
    
        return response.data
    },

    // Update document metadata
    updateDocumentMetadata: async (id, title, description) => {
        const formData = new FormData()
        formData.append("title", title)
    
        if (description) {
            formData.append("description", description)
        }
    
        const response = await api.put(`${DOCUMENTS_URL}/${id}`, formData)
        return response.data
    },  

    // Update document status
    updateDocumentStatus: async (id, status, comments) => {
        const params = { status }
    
        if (comments) {
            params.comments = comments
        }
    
        const response = await api.put(`${DOCUMENTS_URL}/${id}/status`, null, {
            params,
        })
    
        return response.data
    },

    // Delete document
    deleteDocument: async (id) => {
        await api.delete(`${DOCUMENTS_URL}/${id}`)
    },
}

